sdk/monitor/Azure.Monitor.Ingestion/src/Models/UploadLogsOptions.cs

```diff
@@ -20,6 +27,61 @@ public class UploadLogsOptions
         /// The max concurrent requests to send to the Azure Monitor service when uploading logs.
         /// <remarks> In the Upload method, this parameter is not used as the batches are uploaded in sequence. For parallel uploads, if this value is not set the default concurrency will be 5. </remarks>
         /// </summary>
-        public int MaxConcurrency { get; set; }
+        public int MaxConcurrency { get; set; } = 5;
+
+        /// <summary>
+        /// test
+        /// </summary>
+        public event SyncAsyncEventHandler<UploadFailedEventArgs> UploadFailedEventHandler;
+
+        /// <summary>
+        /// test
+        /// </summary>
+        /// <param name="uploadFailedArgs"></param>
+        /// <returns></returns>
+        internal virtual async Task InvokeEvent(UploadFailedEventArgs uploadFailedArgs)
+        {
+            await UploadFailedEventHandler.RaiseAsync(uploadFailedArgs, nameof(LogsIngestionClient), "Upload", uploadFailedArgs._clientDiagnostics).ConfigureAwait(false);
+        }
+
+        /// <summary>
+        /// test
+        /// </summary>
+        /// <param name="eventArgs"></param>
+        internal virtual async Task<Exception> OnUploadFailedAsync(UploadFailedEventArgs eventArgs)
+        {
+            try
+            {
+                if (eventArgs.IsRunningSynchronously)
+                {
+#pragma warning disable AZC0103 // Do not wait synchronously in asynchronous scope.
+                    // for customer code so async not ran over sync
+                    InvokeEvent(eventArgs).GetAwaiter().GetResult();
+#pragma warning restore AZC0103 // Do not wait synchronously in asynchronous scope.
+                }
+                else
+                {
+                    await InvokeEvent(eventArgs).ConfigureAwait(false);
+                }
+                return null;
+            }
+            catch (Exception ex)
+            {
+                // return exception to caller and caller should check exception to abort processing and rethrow this exception
+                return ex;
+            }
+        }
+
+        internal static int AssertNotNegative(int argumentValue, string argumentName)
```

Comment:
We don't seem to be using this.

Reply:
used here: https://github.com/Azure/azure-sdk-for-net/pull/33490/files#diff-6767d7871ae91cf5eb10fa3e4f24a9d5bbd9f9dbcc22bcbd72be0eabe315c103R330 Line 330 in LogsIngestionClient

Reply:
Ah - any reason not to call this in the property setter?